activesupport/lib/active_support/testing/instrumentation_assertions.rb

```diff
@@ -0,0 +1,80 @@
+
+# typed: true
+# frozen_string_literal: true
+
+module ActiveSupport
+  module Testing
+    module InstrumentationAssertions
+      # Assert an event was emitted with a given +pattern+ and optional +payload+.
+      #
+      # You can assert that an event was emitted by passing a pattern, which accepts
+      # either a string or regexp, an optional payload, and a block. While the block
+      # is executed, if a matching event is emitted, the assertion will pass.
+      #
+      #     assert_instrumentation_event("post.submitted", title: "Cool Post") do
+      #       post.submit(title: "Cool Post") # => emits matching ActiveSupport::Notifications::Event
+      #     end
+      #
+      def assert_instrumentation_event(pattern, payload = nil, &block)
+        events = capture_instrumentation_events(pattern, &block)
+        assert_not_empty(events, "No #{pattern} events were found.")
+
+        return if payload.nil?
+
+        event = events.find { |event| event.payload == payload }
+        assert_not_nil(
+          event,
+          "No #{pattern} event with payload #{payload} was found. " \
+            "Did you mean one of these payloads: #{events.map(&:payload).join("\n")}?",
```

Comment:
I got confused and mistook `payload` for `pattern`.
I guess scenarios in which there will be a lot of events on the right pattern but mismatched payload are gonna be super rare, so canceling my previous comment.

 
<details><summary>(I'll leave the content here if you're curious, but please ignore.)</summary>

Could this list become way too big to be reasonably read?
Could joining with `\n` cause a _scroll bomb_?

I was going to suggest the use of `DidYouMean`, but it doesn't look like the Rails codebase is using it much.
<details><summary>For the record, this is what it could look like.</summary>

```rb
error_message = "No #{pattern} event with payload #{payload} was found."
corrections = DidYouMean::SpellChecker.new(dictionary: events.map(&:payload)).correct(pattern)
if corrections.any?
  error_message += " Did you mean one of these payloads: #{corrections.join(", ")}?"
end
assert_not_nil(event, corrections)
```

</details> 

</details> 

Is it actually useful to display other payloads in the error message?

I can imagine multiple payload hashes with multiple keys in each of them could disrupt the flow of error messages, and am unsure this is actually the level of error messaging we need.

A developer could instead easily add a `capture_instrumentation_events` and debug to see what they're missing.

Reply:
Fair points all around, especially about having `capture_instrumentation_events` available for debugging. Originally wrote this assertion helper with this `payload` error message before `capture_instrumentation_events` existed. I think now since that's around we can simplify this whole thing.